domain/secret/state/state.go

```diff
@@ -3154,9 +3155,65 @@ func (st State) DeleteSecret(ctx context.Context, uri *coresecrets.URI, revs []i
 	if err != nil {
 		return errors.Trace(err)
 	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return st.deleteSecretWithRevisions(ctx, tx, uri, revs)
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
 
-	// First delete the specified revisions.
+// DeleteObsoleteUserSecretRevisions deletes the obsolete user secret revisions.
+func (st State) DeleteObsoleteUserSecretRevisions(ctx context.Context) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+SELECT smo.secret_id AS &secretID.id,
+       sr.revision AS &secretExternalRevision.revision
+FROM   secret_model_owner smo
+       JOIN secret_metadata sm ON sm.secret_id = smo.secret_id
+       JOIN secret_revision sr ON sr.secret_id = smo.secret_id
+       LEFT JOIN secret_revision_obsolete sro ON sro.revision_uuid = sr.uuid
+WHERE  sm.auto_prune = true AND sro.obsolete = true`
+
+	stmt, err := st.Prepare(q, secretID{}, secretExternalRevision{}, revisions{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var (
+			dbSecrets    secretIDs
+			dbsecretRevs secretExternalRevisions
+		)
+		err = tx.Query(ctx, stmt).GetAll(&dbSecrets, &dbsecretRevs)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// Nothing to delete.
+			return nil
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+		itemsToDelete, err := dbSecrets.toSecretMetadataForDrain(dbsecretRevs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, toDelete := range itemsToDelete {
+			revs := make([]int, len(toDelete.Revisions))
+			for i, r := range toDelete.Revisions {
+				revs[i] = r.Revision
+			}
+			if err := st.deleteSecretWithRevisions(ctx, tx, toDelete.URI, revs); err != nil {
```

Comment:
This seems to reflect existing behaviour, but I wonder if it's what we want. The method has been refactored from existing DeleteSecret code. If all remaining revisions are removed, then the secret itself is removed. But for pruning obsolete revisions, we are only wanting to remove the revisions that are obsolete. If the entire secret is deleted, this  could be surpising, and removes the possibility of the user being able to add a new revision later. Something to discuss moving forward.

Reply:
The assumption should never happen, because it's not possible to have a secret having all its revisions get obsolete. At least, the latest revision should always be non-obsolete.

Reply:
What about the case where all the consumers of a user secret are removed. Then all the revisions would be obsolete (we should have a test case for this). And I don't think we want to delete the whole secret in that case.

Reply:
I think the latest revision should never become obsolete.